### DISCUSSION TITLE
How to properly eager load column parameter in SQLAlchemy?

### DISCUSSION THREAD

**Discussion by ghevge** [sqlalchemy/sqlalchemy#5997] [Usage Questions]
**Describe your question**
In my tornado webapp, I use SQLAlchemy, and I'm trying to eagerly load an Object, without much success.
SQLAlchemy==1.3.20 tornado==6.0.4

My object structure looks something like this:

`class User(Base):

    __tablename__ = 'users'

    id = Column(Integer, primary_key=True)
    first_name = Column(String(100), nullable=True)
    last_name = Column(String(100), nullable=True)
    name = column_property(first_name + ' ' + last_name)
    telephone_num = Column(String(50), nullable=True)
    password = Column(String(500), nullable=True)
    password_salt = Column(String(100), nullable=True)
    email = Column(String(255), nullable=False, unique=True)
    created_at = Column(DateTime, default=datetime.datetime.utcnow)
    last_logged_in = Column(DateTime, nullable=True)
    activated_at = Column(DateTime, nullable=True)
    last_updated = Column(DateTime, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow)
    language = Column(String(10), default='en')
    api_key = Column(String(45), default=generate_api_key, unique=True)
    is_enabled = Column(Boolean, nullable=False, default=True)
    is_activated = Column(Boolean, nullable=False, default=False)
    last_viewed_release = Column(String(10), nullable=True, default=None)
    company = Column(String(100), nullable=True, default=None)
    oauth_user_id = Column(String(100), nullable=True, default=None)
    oauth_clients = relationship('UserOauthClient',
                                 foreign_keys='UserOauthClient.user_id',
                                 cascade='all, delete-orphan')

    notices_read = relationship(
        'Notice',
        secondary='user_notice_associations',
        passive_deletes=True,
    )
    namespaces = relationship(  # Lazy-loaded, queriable
        'Namespace',
        lazy='dynamic',
        secondary='user_namespace_roles',
        backref='users',
        cascade='save-update, merge',
        passive_deletes=True,
        foreign_keys='[UserNamespaceRole.user_id, UserNamespaceRole.namespace_id]',
    )
    all_namespaces = relationship(  # Eager-loadable, non-queriable
        'Namespace',
        secondary='user_namespace_roles',
        cascade='save-update, merge',
        passive_deletes=True,
        foreign_keys='[UserNamespaceRole.user_id, UserNamespaceRole.namespace_id]',
    )

    @property
    def projects(self):
        return self.find_all_projects()

    @property
    def applications(self):
        return self.find_all_apps()

    @property
    def is_internal(self):
        return self.namespaces.filter_by(path=Namespace.NUANCE_PATH).first() is not None
......`

I'm fetching the User object like this:

```
user_q = self.application.db.query(User).filter(User.id == request.user_id)

user_q.options(undefer('*'), joinedload('*'))

fetched_user = user_q.one_or_none()
```

And then I store this object into an ExpiringDict. Basically I'm trying to set up a cache, in order to fix some connection pool exhaustion problems. For the request that sets this object, everything is ok, but for subsequent requests I get errors like:

`Traceback (most recent call last):
mix-middleware-server    |   File "/bolt-webserver/bolt/api/handlers.py", line 294, in user
mix-middleware-server    |     if not user.is_enabled:
mix-middleware-server    |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py", line 287, in __get__
mix-middleware-server    |     return self.impl.get(instance_state(instance), dict_)
mix-middleware-server    |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py", line 718, in get
mix-middleware-server    |     value = state._load_expired(state, passive)
mix-middleware-server    |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/state.py", line 652, in _load_expired
mix-middleware-server    |     self.manager.deferred_scalar_loader(self, toload)
mix-middleware-server    |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/loading.py", line 942, in load_scalar_attributes
mix-middleware-server    |     raise orm_exc.DetachedInstanceError(
mix-middleware-server    | sqlalchemy.orm.exc.DetachedInstanceError: Instance <User at 0x7f9191ba28e0> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: http://sqlalche.me/e/13/bhk3)`

So definitely the User object is not eagerly loaded. Any idea how I can force SQLAlchemy to completely load that object when I ask for it, and don't keep any session dependencies around?

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**

**Complete stack trace, if applicable**

**Versions**
 - OS: Linux
 - Python: 3.8
 - SQLAlchemy: 1.3.20
 - Tornado: 6.0.4
 - Database: PostgreSQL
 - DBAPI:

**Additional context**

**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
it looks like your User object is getting expired.    The "user.is_enabled" is a simple column on that User object, so undefer or joinedload has no impact on this.    Make sure you properly [detach](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.expunge) those objects from the Session before the Session runs through a commit() or rollback() that would expire everything.

**Reply by ghevge** [reply]
Thanks, but how do I force SQLAlchemy to eager load all the elements referenced by my object, before detaching it ?

**Reply by zzzeek** [reply]
the `joinedload("*")` will do that but it might produce an inefficient query.   `selectinload("*")` probably best.  use echo=True on your engine to verify you see the SQL emitted that you expect.
